Document the NO_OP TariScript vulnerability

philipr-za opened the pull request:
## Description
This PR contains a write up off the NO_OP TariScript vulnerability spotted by @mikethetike .

This was put into the mdbook in the repo in /docs/. To view the document do the following:

In a terminal navigate to /docs
cargo install mdbook
mdbook serve
In a browser navigate to localhost:3000

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [x] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [ ] I ran `cargo-fmt --all` before pushing.
* [ ] I ran `cargo test` successfully before submitting my PR.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [x] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

SWvheerden commented on docs/src/tari_script_no_op_vulnerability.md:
```diff
@@ -0,0 +1,82 @@
+
+The initial specification of how TariScript is incorporated into a Tari Mimblewimble (MW) style transaction is defined in the Tari [RFC-0201](https://rfc.tari.com/RFC-0201_TariScript.html). 
+
+This vulnerability applies to transactions where a Sender is spending one of their inputs that is bound by a NO_OP script like is suggested in [RFC-0201](https://rfc.tari.com/RFC-0201_TariScript.html#standard-mw-transaction)
```
```suggestion
This vulnerability applies to transactions where a Sender is spending one or more their inputs where all the inputs are bound by a NO_OP script like is suggested in [RFC-0201](https://rfc.tari.com/RFC-0201_TariScript.html#standard-mw-transaction)
```

SWvheerden replied:
This is only a problem where all the inputs and all the outputs are changeable. 

if 
\\( \gamma = k_{sa1} + k_{sa2} - k_{Ob}U_b  \\)
Then Bob cant construct a k_{Ob} to satisfy (k_{sa1} + k_{sa2})

philipr-za replied:
Sure updated the wording

SWvheerden commented on docs/src/tari_script_no_op_vulnerability.md:
```diff
@@ -0,0 +1,82 @@
+
+\\( \hat{\gamma} =  \hat{k_{sa}} - \hat{k_{Ob}} \hat{U_b} \\).
+
+This attack is only possible if Alice did not have a change output in the transaction because in that case Bob would not know the private script offset for the change output, \\( k_{oc} \\), to produce 
```
```suggestion
This attack is only possible if Alice did not have a change output that he does not know the blinding factor for in the transaction because in that case Bob would not know the private script offset for the change output, \\( k_{oc} \\), to produce 
```

This attack is only open are ALL inputs are NO_OP and Bob knows the blingfactor for ALL outputs.

philipr-za replied:
lol, I am not sure how to measure the `bling` of an output. 

But I will update the wording to include the idea that all the inputs need to be NO_OP and the change output. I will briefly mention that its all outputs but I prefer to keep the cleaner paradigm of 1 change output and 1 receiver output are the outputs that can be in a Tx.

SWvheerden reviewed (commented):
Looks good, just ttwo comments

philipr-za committed:
Document the NO_OP TariScript vulnerability	

This PR contains a write up off the NO_OP TariScript vulnerability spotted by @mikethetike.

SWvheerden reviewed (approved):
LGTM
